[BUG] Fix pivot index calculations for naive forecaster

benHeid opened the pull request:
<!--
Welcome to sktime, and thanks for contributing!
Please have a look at our contribution guide:
https://www.sktime.net/en/latest/get_involved/contributing.html
-->

#### Reference Issues/PRs
Fix #5032 

#### What does this implement/fix? Explain your changes.
Fixes issues with casting of PeriodIndex using astype("int64"), since this casting is probably using the frequency to determine the distances between two consecutive steps. 

#### Does your contribution introduce a new dependency? If yes, which one?

 Noe
#### What should a reviewer concentrate their feedback on?

Does this change makes sense. Is this affecting also other Estimators? 

#### Did you add any tests for the change?

No
#### Any other comments?


#### PR checklist
<!--
Please go through the checklist below. Please feel free to remove points if they are not applicable.
-->

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)

- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.


<!--
Thanks for contributing!
-->

fkiraly reviewed (changes requested):
I think the issue is, `y_pred` ends up being `Index` instead of a specific type of index. Not sure why, `pandas` is often strange like this.

fkiraly reviewed (dismissed):
I see, so the failure condition is for multiples.

benHeid commented on sktime/forecasting/tests/test_naive.py:
```diff
@@ -401,3 +401,18 @@ def test_naive_predict_interval_against_R_naive(strategy, sp, lower, upper):
+
+
+def test_naive_sp_greater_1_not_nan():
+    sample_dates = pd.date_range(start="2001-01-01", periods=30, freq="2D")
```
Thank you. I tested it and it didn't worked with your specified indices.

fkiraly replied:
is this the failing case?

yarnabrina replied:
Thanks for checking the bad cases, I continuously face issues with these frequencies at work, so guessed as such.

If it's okay with you, will you please parametrise the test for these frequencies as well? I think adding something like this and taking the `frequency` as argument to the test function will be enough.

```python
@pytest.mark.parametrize("frequency", ["D", "W", "W-TUE", "MS", "ME"])
```

benHeid replied:
I added the test with parameters. But I think "ME" is no valid frequency. And "MS" leads to an exception that is not related to this issue.

yarnabrina replied:
> I think "ME" is no valid frequency

I can see `ME` here though. May be it's because of pandas version.

https://pandas.pydata.org/docs/user_guide/timeseries.html#offset-aliases

---

Another small request, and I don't know if it matters or not, can you please add `2W` along with `W` and etc. in the tests? This is irrespective of the final fix you and @fkiraly decides.

(I am not able to follow this discussion becuase of so many problems and solutions you are trying.)

---

Is something like this useful in any way?

```python
>>>
>>> a = pd.Timestamp(year=2024, month=1, day=31)
>>>
>>> a + pd.offsets.MonthEnd(n=1)
Timestamp('2024-02-29 00:00:00')
>>>
>>> a + pd.offsets.MonthEnd(n=-2)
Timestamp('2023-11-30 00:00:00')
>>>
```

(I am sorry if it's completely unrelated, I am not familiar with how exactly forecasting horizon works and it seems too complicated to learn quickly.)

benHeid commented:
I changed how the NaiveForecaster works in this scenario, since I believe that the combination of the specified frequencies with sp>2 is not working with pivot_tables. At least I haven't found a solution. Thus, I implemented an alternative that is built upon the grouping functionality of pandas. I hope reading that code is not much harder than the old code. Feel free to ask for clarification, since I am unsure if my comments are sufficient.

fkiraly reviewed (changes requested):
Interesting. Some questions:

* have you checked performance? "numpy done right" tends to outperform `pandas`
* could you kindy take a look at this more general reducer, and see if your alternative would make it easier to complete or not? https://github.com/sktime/sktime/pull/4500
* can you explain to me what is broken on the eve of pivot/unpivot?

benHeid commented:
> Interesting. Some questions:
> 
> * have you checked performance? "numpy done right" tends to outperform `pandas`
I haven't checked it. But the previous solution was also very pandas focussed I think. But a comparison of the performance would be good. 

> * could you kindy take a look at this more general reducer, and see if your alternative would make it easier to complete or not? [[ENH] seasonal forecast wrapper to use any forecaster as seasonal forecaster #4500](https://github.com/sktime/sktime/pull/4500)


> * can you explain to me what is broken on the eve of pivot/unpivot?

The problem with the pivot/unpivot was/is its internal combination with the PeriodIindex. The resulting index for a weekly time series after the pivot was:

```
01.01.2020/08.01.2020, ....
14.01.2020/21.01.2020, ....
```

After the unpivot the forecast index were in some cases one day off. And I haven't seen an easy solution to fix it. If you are interested, I can push the other solution with that issue on a do-not-merge branch. Perhaps it could also be solved with the pivot, but I am more familiar with the grouping operation. Thus, I implemented it with grouping.

fkiraly commented:
> If you are interested, I can push the other solution with that issue on a do-not-merge branch. Perhaps it could also be solved with the pivot, but I am more familiar with the grouping operation. Thus, I implemented it with grouping.

That would be appreciated!

This functionality is aso quite heavily tested, which is why I was surprised in the first place that there were bugs - if it if not too much of a hassle, could you also push a PR that contains only a test?

fkiraly commented:
it would be nice if you could have a quick look at `test_pivot_sp` and check where we missed coverage of this. Is it only period frequencies that are non-trivial multiples?

benHeid commented:
> > If you are interested, I can push the other solution with that issue on a do-not-merge branch. Perhaps it could also be solved with the pivot, but I am more familiar with the grouping operation. Thus, I implemented it with grouping.
> 
> That would be appreciated!

See #5923

> This functionality is aso quite heavily tested, which is why I was surprised in the first place that there were bugs - if it if not too much of a hassle, could you also push a PR that contains only a test?
> 
> update: I see there is a test included, so no need to do that.

I added further parameters for the frequencies.

benHeid commented:
> test_pivot_sp

I think the tests cover only specific frequencies. Thus, if you would use the following df, it leads to an assertion error:

```python

sample_dates = pd.date_range(start="2001-01-01", periods=30, freq="W")
sample_values = np.random.default_rng(seed=0).random(size=len(sample_dates))

df = pd.Series(sample_values, index=sample_dates)
```

fkiraly reviewed (changes requested):
I have looked into this more now - a concern is the reduction of test coverage - previously this was by `test_naive` and `test_seasonal`, but the latter no longer applies.

In particular, I am worried that the core logic is no longer covered for diverse index combinations. Could you either add this, or explain why it is fine?

benHeid commented:
> Other silly question: if you no longer import `_pivot_sp` and `_unpivot_sp`, why are you changing them?

If we decide, to go without pivot, I will remove these changes as mentioned here: https://github.com/sktime/sktime/pull/5891#issuecomment-1936576761

fkiraly commented:
I see - can you then bring this PR to its "final" state? #5923 has the other fixes

yarnabrina commented:
I can see it here though. May be it's because of pandas version.

https://pandas.pydata.org/docs/user_guide/timeseries.html#offset-aliases

Another small request, and I don't know if it matters or not, can you
please add 2W along with W and etc.?

On Sat, Feb 10, 2024, 16:42 Benedikt Heidrich ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In sktime/forecasting/tests/test_naive.py
> <https://github.com/sktime/sktime/pull/5891#discussion_r1485061771>:
>
> > @@ -401,3 +401,18 @@ def test_naive_predict_interval_against_R_naive(strategy, sp, lower, upper):
>      expected[(0, coverage, "upper")] = upper
>
>      pd.testing.assert_frame_equal(y_pred_ints, expected)
> +
> +
> +def test_naive_sp_greater_1_not_nan():
> +    sample_dates = pd.date_range(start="2001-01-01", periods=30, freq="2D")
>
> I added the test with parameters. But I think "ME" is no valid frequency.
> And "MS" leads to an exception that is not related to this issue.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/sktime/sktime/pull/5891#discussion_r1485061771>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJMCQBCBX4T4LZ66FZYGFELYS5IY7AVCNFSM6AAAAABCZC45JGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTQNZTGYYTMOBUGQ>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

benHeid commented:
> I see - can you then bring this PR to its "final" state? #5923 has the other fixes

Yes I will finalize it in the next day and ping if it is ready. #5923 is an alternative version that focuses on the `_sp_pivot`. However, I am not sure if this fix is then still needed, since I have no overview where the pivot functionality is used.

benHeid commented:
That is no fast fix I suppose. Frequencies like `M`, will lead to problems since the period_len_int varies between different months.

fkiraly commented:
> However, I am not sure if this fix is then still needed, since I have no overview where the pivot functionality is used.

I'll try to fix it there as well - I was using it in a (stalled) rework of the `NaiveForecaster` - it has lots of performance problems, see
https://github.com/sktime/sktime/issues/4460

it is one of the most frequently used estimators, and from a methods perspective it is very odd. I started with the most common strategy, but the others still have the problem - and for `sp>1` I was going to use the pivot functionaliy, since it seems easier to generalize.

fkiraly commented:
superseded by https://github.com/sktime/sktime/pull/5923, I think
